fix: Load `SalesChannelCategoryEntity` when loading the breadcrumb to properly access the seo links

aragon999 opened the pull request:
### 1. Why is this change necessary?
Otherwise the SEO URLs in the category breadcrumb are missing:
![image](https://github.com/user-attachments/assets/f704da90-f017-4e16-bdb2-7c995b2cf6bc)

### 2. What does this change do, exactly?
Use the sales channel repository to load the categories, if the sales channel context can be accessed. This will be a hard requirement for Shopware 6.8.

### 3. Describe each step to reproduce the issue or behaviour.
Check the links of the breadcrumb.

### 4. Please link to the relevant issues (if any).
\-

### 5. Checklist

- [x] I have written tests and verified that they fail without my change
- [x] I have created a [changelog file](https://github.com/shopware/shopware/blob/trunk/adr/2020-08-03-implement-new-changelog.md) with all necessary information about my changes
- [ ] I have written or adjusted the documentation according to my changes
- [ ] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfill them.

aragon999 committed:
fix: Load `SalesChannelCategoryEntity` when loading the breadcrumb

to properly access the seo links

mitelg commented on src/Core/Framework/Adapter/Twig/Extension/BuildBreadcrumbExtension.php:
```diff
@@ -19,35 +19,28 @@ class BuildBreadcrumbExtension extends AbstractExtension
+     * @return array<string, SalesChannelCategoryEntity>
      */
-    public function getFullBreadcrumb(array $twigContext, CategoryEntity $category, Context $context): array
+    public function getFullBreadcrumb(CategoryEntity $category, SalesChannelContext $context): array
```
that is definitely a break. and we do only breaking changes until the first RC :slightly_smiling_face: 

so, you need to change it back unfortunately. but other than that, the change looks okay for me on first sight

aragon999 replied:
@mitelg I changed it, to be not a break anymore. At least I think so. 
The BC Checker complains about `CHANGED: The parameter $context of Shopware\Core\Framework\Adapter\Twig\Extension\BuildBreadcrumbExtension#getFullBreadcrumbById() changed from Shopware\Core\Framework\Context to Shopware\Core\Framework\Context|Shopware\Core\System\SalesChannel\SalesChannelContext`

But as I still can pass the old argument, I think it should be okay, right? Or do I miss something.

mitelg replied:
Imo, extending the allowed type is not break, as calls of the method would not break with this change. 

I guess the tool is just not smart enough to detect this :grin: 

so, you can add it to the ignore config

aragon999 replied:
Perfect :)

aragon999 committed:
fixup: Fix break

aragon999 committed:
fixup: Changelog linting

aragon999 committed:
fixup: Fix deprecation notice

mitelg commented on changelog/_unreleased/2025-05-10-load-saleschannelcategoryentity-when-loading-the-breadcrumb.md:
```diff
@@ -0,0 +1,19 @@
+* Changed the Twig functions `sw_breadcrumb_full()` and `sw_breadcrumb_full_by_id()` to load `SalesChannelCategoryEntity` in order to properly display the corresponding seo link in the breadcrumb
+___
+# Upgrade Information
+change in the Twig templates
```
this needs a `##` headline, so it could be merged with the other changelogs. 

everything under `# Upgrade Information` will be merged together, so it would look strange, that this flies around somewhere

aragon999 replied:
@mitelg mh, I added the information to the `UPGRADE-6.8.md` file. Is this then not required? Then I would replace the information from there in this changelog.

mitelg replied:
this changelog file should contain how to upgrade now, to prevent the deprecation. the 6.8 upgrade guide should document what will be removed or changed and how to fix that, if not already done

mitelg replied:
yes, it could be kind of duplicated, but people will not remember how to fix that with 6.8, just they read it here on some 6.7 release :wink:

aragon999 replied:
Okay, not sure, but is it is now in the PR correct? Or did I misunderstood anything? :)

mitelg replied:
is okay now :+1:

aragon999 replied:
Perfect :-)

mitelg commented on src/Core/Framework/Adapter/Twig/Extension/BuildBreadcrumbExtension.php:
```diff
@@ -59,7 +73,12 @@ public function getFullBreadcrumb(array $twigContext, CategoryEntity $category,
         $criteria->setTitle('breadcrumb-extension');
-        $categories = $this->categoryRepository->search($criteria, $context)->getEntities();
+
+        if ($context instanceof SalesChannelContext) {
```
we should surround that if/else with a feature flag, so we can remove the deprecated parts easier with 6.8. 

```
if (feature(6.8) is active) {
$categories = $this->salesChannelCategoryRepository->search($criteria, $context)->getEntities();
} else {
 if ($context instanceof SalesChannelContext) {
            $categories = $this->salesChannelCategoryRepository->search($criteria, $context)->getEntities();
        } else {
            $categories = $this->categoryRepository->search($criteria, $context)->getEntities();
}
}

```

mitelg commented on src/Core/Framework/Adapter/Twig/Extension/BuildBreadcrumbExtension.php:
```diff
@@ -38,16 +45,23 @@ public function getFunctions(): array
     /**
      * @param array<string, mixed> $twigContext
```
this parameter should also be deprecated, as it will be removed, right?

mitelg commented on src/Core/Framework/Adapter/Twig/Extension/BuildBreadcrumbExtension.php:
```diff
@@ -76,15 +95,47 @@ public function getFullBreadcrumb(array $twigContext, CategoryEntity $category,
+            $context = $this->getSalesChannelContext($twigContext) ?? $context;
+        }
+
+        if ($context instanceof SalesChannelContext) {
```
same as above

mitelg commented on src/Core/Framework/Adapter/Twig/Extension/BuildBreadcrumbExtension.php:
```diff
@@ -76,15 +95,47 @@ public function getFullBreadcrumb(array $twigContext, CategoryEntity $category,
+
+        Feature::triggerDeprecationOrThrow(
+            'v6.8.0.0',
+            'SalesChannelContext is required in v6.8.0.0.'
```
is it clear, where this deprecation is triggered just from this message? :thinking:

aragon999 replied:
I think with the changes we can just remove it at this place.

aragon999 committed:
fixup: Exclude type extension from BC check

aragon999 committed:
fixup: Implement review changes

aragon999 committed:
fixup: Add deprecations

aragon999 commented:
I think now I resolved the issues, and marked the PR as not draft anymore.

mitelg commented on src/Core/Framework/Adapter/Twig/Extension/BuildBreadcrumbExtension.php:
```diff
@@ -19,35 +23,50 @@ class BuildBreadcrumbExtension extends AbstractExtension
-        $salesChannel = null;
-        if (\array_key_exists('context', $twigContext) && $twigContext['context'] instanceof SalesChannelContext) {
-            $salesChannel = $twigContext['context']->getSalesChannel();
+        if (Feature::isActive('v6.8.0.0')) {
```
the feature flag usage here is not correct. 

this would mean, the code below would only be executed, if the flag is active. but then the deprecation would never be triggered in default 6.7 environments. 

you should use the feature active check with a negation, and execute the old behaviour within this condition. then with 6.8, we can completely remove this code block and have the new behaviour without much thinking about what could removed or what need to stay. 
it is okay, if this means, we have two big branches in if/else containing much similar code.

mitelg commented on src/Core/Framework/Adapter/Twig/Extension/BuildBreadcrumbExtension.php:
```diff
@@ -74,17 +98,53 @@ public function getFullBreadcrumb(array $twigContext, CategoryEntity $category,
     {
-        $category = $this->categoryRepository->search(new Criteria([$categoryId]), $context)->getEntities()->first();
-        if (!$category instanceof CategoryEntity) {
+        if (Feature::isActive('v6.8.0.0')) {
```
same as above.

aragon999 replied:
Argh, I forgot to commit the proper changes :see_no_evil: for the `getFullBreadcrumb` method...

I meant it to be like it is here, i.e. if the Feature `v6.8.0.0` is active, we can assert that the context is an instance of the `SalesChannelContext` and directly use the `salesChannelCategoryRepository`, otherwise we try to determine the proper context, is this okay? Or should I turn the conditions around? :thinking:

aragon999 committed:
fixup: Correct getFullBreadcrumb() method with deprecations

mitelg commented on src/Core/Framework/Adapter/Twig/Extension/BuildBreadcrumbExtension.php:
```diff
@@ -19,35 +23,55 @@ class BuildBreadcrumbExtension extends AbstractExtension
     }
 
     /**
+     * @deprecated tag:6.8.0 - Parameter $twigContext will be removed, as it is not needed anymore and the type of `$context` will be changed to `SalesChannelContext`
```
```suggestion
     * @deprecated tag:v6.8.0 - Parameter $twigContext will be removed, as it is not needed anymore and the type of `$context` will be changed to `SalesChannelContext`
```

mitelg replied:
```suggestion
     * @deprecated tag:v6.8.0 - Parameter $twigContext will be removed, as it is not needed anymore and the type of `$context` will be changed to `SalesChannelContext`
```

mitelg commented on src/Core/Framework/Adapter/Twig/Extension/BuildBreadcrumbExtension.php:
```diff
@@ -19,35 +23,55 @@ class BuildBreadcrumbExtension extends AbstractExtension
 
     /**
+     * @deprecated tag:6.8.0 - Parameter $twigContext will be removed, as it is not needed anymore and the type of `$context` will be changed to `SalesChannelContext`
+     * @deprecated tag:6.8.0 - reason:return-type-change - Will only return `array<string, SalesChannelCategoryEntity>`
```
```suggestion
     * @deprecated tag:v6.8.0 - reason:return-type-change - Will only return `array<string, SalesChannelCategoryEntity>`
```

mitelg replied:
```suggestion
     * @deprecated tag:v6.8.0 - reason:return-type-change - Will only return `array<string, SalesChannelCategoryEntity>`
```

mitelg commented on src/Core/Framework/Adapter/Twig/Extension/BuildBreadcrumbExtension.php:
```diff
@@ -74,17 +109,53 @@ public function getFullBreadcrumb(array $twigContext, CategoryEntity $category,
     }
 
     /**
+     * @deprecated tag:6.8.0 - Parameter $twigContext will be removed, as it is not needed anymore and the type of `$context` will be changed to `SalesChannelContext`
```
```suggestion
     * @deprecated tag:v6.8.0 - Parameter $twigContext will be removed, as it is not needed anymore and the type of `$context` will be changed to `SalesChannelContext`
```

mitelg replied:
```suggestion
     * @deprecated tag:v6.8.0 - Parameter $twigContext will be removed, as it is not needed anymore and the type of `$context` will be changed to `SalesChannelContext`
```

mitelg commented on src/Core/Framework/Adapter/Twig/Extension/BuildBreadcrumbExtension.php:
```diff
@@ -74,17 +109,53 @@ public function getFullBreadcrumb(array $twigContext, CategoryEntity $category,
 
     /**
+     * @deprecated tag:6.8.0 - Parameter $twigContext will be removed, as it is not needed anymore and the type of `$context` will be changed to `SalesChannelContext`
+     * @deprecated tag:6.8.0 - reason:return-type-change - Will only return `array<string, SalesChannelCategoryEntity>`
```
```suggestion
     * @deprecated tag:v6.8.0 - reason:return-type-change - Will only return `array<string, SalesChannelCategoryEntity>`
```

mitelg replied:
```suggestion
     * @deprecated tag:v6.8.0 - reason:return-type-change - Will only return `array<string, SalesChannelCategoryEntity>`
```

mitelg commented on changelog/_unreleased/2025-05-10-load-saleschannelcategoryentity-when-loading-the-breadcrumb.md:
```diff
@@ -0,0 +1,19 @@
+# Upgrade Information
+## Breadcrumb template functions require the `SalesChannelContext`
+
+The Twig breadcrumb functions `sw_breadcrumb_full` and `sw_breadcrumb_full_by_id` now require the `SalesChannelContext`, i.e. adjust the default Twig templates as follows
```
```suggestion
The Twig breadcrumb functions `sw_breadcrumb_full` and `sw_breadcrumb_full_by_id` will require the `SalesChannelContext` with the next major version.
Adjust the Twig templates as follows
```

BrocksiNet commented:
Hi @aragon999 and thanks for that PR 🎉 

In my cloud test instance that is based on 6.7 RC 5, I got this output:
![image](https://github.com/user-attachments/assets/c9bf1c19-eb4a-4621-982b-fbc7f1e73e86)

So my question is 😀 

You updated the unit tests to work with `SalesChannelCategoryEntity`. But I wonder if we are missing some tests to cover this case where the href is empty. As you can see on my screen, it is filled.

So, when is it empty? Depends on some config or setup?

keulinho commented on src/Core/Framework/Adapter/Twig/Extension/BuildBreadcrumbExtension.php:
```diff
@@ -19,35 +23,55 @@ class BuildBreadcrumbExtension extends AbstractExtension
+     * @return array<string, CategoryEntity|SalesChannelCategoryEntity>
      */
-    public function getFullBreadcrumb(array $twigContext, CategoryEntity $category, Context $context): array
+    public function getFullBreadcrumb(array $twigContext, CategoryEntity $category, Context|SalesChannelContext $context): array
```
we can get the context from the twig context, no need to pass it explicitly and with that it should not be a break ;) 
```php
        if (\array_key_exists('context', $twigContext) && $twigContext['context'] instanceof SalesChannelContext) {
            $salesChannel = $twigContext['context']->getSalesChannel();
        }
        ```
        
        That's from the now deprecated CategoryUrlExtension

keulinho replied:
ah you use it already, but can't we just ignore the second parameter then and make it nullable

aragon999 replied:
@keulinho yes, but what should be done if it does not exist in the twig context? And with the change, the `Context` would not needed anymore.

So I think it might be better to explicitly add the context in the template, because otherwise the error might be very hidden (if the context is missing).

The sales channel is optional in the subsequent function calls.

For example, what happens if the breadcrumb template is included with the `only` keyword?

keulinho replied:
i don't know how the twig context thingy is working under the hood, I was just wondering why we didn't pass it in the other twig function that needs it, your explanation makes sense, we should pass it directly

aragon999 commented:
@BrocksiNet the change which introduced the problem is currently only in `trunk`: https://github.com/shopware/shopware/commit/f88b6769984086bd73b6cb8946d9ce270d04d249 :-)

keulinho commented:
@BrocksiNet keep in mind the base feature that requires the sales channel entity is only in trunk and not in 6.7.0.0RC

mitelg commented:
but a test, that verifies the change, would be nice nevertheless :wink:

mitelg commented:
and I guess your PR description needs a change after the code changes you did :slightly_smiling_face:

aragon999 commented:
> but a test, that verifies the change, would be nice nevertheless 😉

Question is what exactly should be tested? The `seoLink` itself is added in the subscriber: https://github.com/shopware/shopware/blob/trunk/src/Core/Content/Category/Subscriber/CategorySubscriber.php#L70

And the type is "verified" by the return type :thinking: 

> and I guess your PR description needs a change after the code changes you did 🙂

Ehm, are you sure? :-) I think the description itself is still valid, as all the changes where only there to make sure everything is backwards compatible?

mitelg commented:
> Ehm, are you sure? :-) I think the description itself is still valid, as all the changes where only there to make sure everything is backwards compatible?

was just about this part :sweat_smile: `Not sure if this a too hard break...`, which it isn't now anymore :wink:

mitelg commented:
> Question is what exactly should be tested?

I guess it would need some kind of an integration test :thinking: not sure, if this would need some twig rendering :grimacing:

aragon999 committed:
fixup: Make sure that we have SalesChannelCategoryEntities in the Breadcrumb build process

aragon999 committed:
fixup: Add integration test for category subscriber

aragon999 commented:
@mitelg I rebased and fixed the merge conflict.

I also added a test case for the `sales_channel.category.loaded` event, that there the correct seo url is set.

Also checked that we get a `SalesChannelCategoryEntity` when we build the breadcrumb.

mitelg committed:
Apply suggestions from code review

mitelg committed:
Merge branch 'trunk' into fix/breadcrumb-category-urls

mitelg commented:
thanks for your contribution @aragon999 👍🎉💙
